processing/src/main/java/org/apache/druid/query/QueryToolChest.java

```diff
@@ -330,4 +330,29 @@ public Sequence<Object[]> resultsAsArrays(QueryType query, Sequence<ResultType>
   {
     throw new UOE("Query type '%s' does not support returning results as arrays", query.getType());
   }
+
+  /**
+   * Converts a sequence of this query's ResultType into a sequence of {@link FrameSignaturePair}. The array signature
+   * is the one give by {@link #resultArraySignature(Query)}.
+   *
+   * Check documentation of {@link #resultsAsArrays(Query, Sequence)} as the behaviour of the rows represented by the
+   * frame sequence is identical.
+   *
+   * Each Frame has a separate {@link RowSignature} because for some query types like the Scan query, every
+   * column in the final result might not be present in the individual ResultType (and subsequently Frame). Therefore,
+   * this is done to preserve the space by not populating the column in that particular Frame and omitting it from its
+   * signature
+   */
+  public Sequence<FrameSignaturePair> resultsAsFrames(QueryType query, Sequence<ResultType> resultSequence)
```

Comment:
Can we return an Optional<Sequence> and remove the `canFetchResultsAsFrames` call from the interface ?

Reply:
Updated